### PR TITLE
HTTP2 pool support

### PR DESCRIPTION
**Pull request by keathley**
I hope you don't mind @whatyouhide but I kinda wanted to give this a try. Mostly because I wanted to make sure I understood exactly what was going on here. This is based on the work that y'all had done in mint_pool. The only real change is the removal of the explicit streaming options.

I've added some very basic tests in here. But we should work on testing more of the failure conditions and request timeouts. I think we're also going to have to work out what to do if we already have too many inflight requests for a given pool. Its possible that we can buffer a small number of requests. But that buffer would have to be bounded. It also means that we would need to look at implementing something like CoDel in order to evict requests from the buffer. I think both of these are reasonable and make sense to add to Finch. For now we're just going to drop the requests, which is a safer option for users.

Lemme know what you think.

### CODE REVIEW

**Comment by whatyouhide** [comment]
> I hope you don't mind @whatyouhide

I don't :)

> The only real change is the removal of the explicit streaming options.

Why?

> Its possible that we can buffer a small number of requests. But that buffer would have to be bounded. It also means that we would need to look at implementing something like CoDel in order to evict requests from the buffer. I think both of these are reasonable and make sense to add to Finch. For now we're just going to drop the requests, which is a safer option for users.

Make it work, make it beautiful 😛 I definitely think this is a "post MVP" feature.


You could take a bunch of integration tests from mint_pool for this. At least they cover _something_ :)

**Comment by keathley** [comment]
@whatyouhide I dropped the explicit streaming function since there's no top-level `stream_request` type API for Finch. We'd need to make sure that that would work with http 1.1 requests as well before rolling it out. We haven't needed to do that yet but we can always investigate in the future. For now the http/2 pool only exposes a `request` function which collects the full response before returning to the user.

**Comment by sneako** [comment]
Really excited to try this out! @keathley how about the telemetry support? Were you thinking that should be added in a later PR?

**Comment by keathley** [comment]
Yup. I’d like to add telemetry and some more tests before we merge this in. We should probably do some benchmarking as well just to see where we stand

**Comment by keathley** [comment]
@sneako Updated with telemetry and more tests.

File: lib/finch/http2/pool.ex [hunk 1]
```diff
@@ -0,0 +1,409 @@
...
+      {:ok, conn} ->
+        Telemetry.stop(:connect, start, metadata)
+        data = %{data | conn: conn}
+        {:next_state, :connected, data}
+
+      {:error, error} ->
+        metadata = Map.put(metadata, :error, error)
+        Telemetry.stop(:connect, start, metadata)
+        Logger.error([
+          "Failed to connect to #{data.scheme}:#{data.host}:#{data.port}: ",
```
Comment (sneako) [inline lib/finch/http2/pool.ex:160]:
```suggestion
          "Failed to connect to #{data.scheme}://#{data.host}:#{data.port}: ",
```

File: lib/finch.ex [hunk 2]
```diff
@@ -147,6 +147,9 @@ defmodule Finch do
         query: uri.query
       }
 
+      # TODO - Use Nimble options to build teh request opts
```
Comment (sneako) [inline lib/finch.ex:150]:
```suggestion
      # TODO - Use Nimble options to build the request opts
```

File: lib/finch/http2/pool.ex [hunk 3]
```diff
@@ -0,0 +1,409 @@
...
+
+        if HTTP2.open?(conn, :read) do
+          {:keep_state, data, actions}
+        else
+          {:next_state, :disconnected, data, actions}
+        end
+
+      {:error, conn, error, responses} ->
+        Logger.error([
+          "Received error from server #{data.scheme}:#{data.host}:#{data.port}: ",
```
Comment (sneako) [inline lib/finch/http2/pool.ex:324]:
```suggestion
          "Received error from server #{data.scheme}://#{data.host}:#{data.port}: ",
```

File: test/finch/http2/pool_test.exs [hunk 4]
```diff
@@ -0,0 +1,281 @@
...
+    {:ok, pool} =
+      start_server_and_connect_with([server_settings: server_settings], fn port ->
+        start_pool(port)
+      end)
+
+    assert {:error, error} = Pool.request(pool, %{req | headers: [{"foo", "bar"}]}, [])
+    assert %{reason: {:max_header_list_size_exceeded, _, _}} = error
+  end
+
+  @tag :capture_log
```
Comment (sneako) [inline test/finch/http2/pool_test.exs:71]:
```suggestion
```
Since we have the `@moduletag`

File: test/finch/http2/pool_test.exs [hunk 5]
```diff
@@ -0,0 +1,281 @@
...
+    # If the server now closes the socket, we actually shut down.
+    :ok = :ssl.close(server_socket())
+
+    Process.sleep(50)
+
+    # If we try to make a request now that the server shut down, we get an error.
+    assert {:error, %{reason: :disconnected}} = Pool.request(pool, req, [])
+  end
+
+  @tag :capture_log
```
Comment (sneako) [inline test/finch/http2/pool_test.exs:108]:
```suggestion
```

**Pull request merged**